vanilla/vanilla
# Restructure PHP Quill renderer for more consistent rendering

Closes https://github.com/vanilla/vanilla/issues/7194

## TL;DR How to test

- Check the `/richeditorstyle/rendered` page.
- Create a post with as much stuff as you can think of, and make sure hitting "preview" or posting it results in the same output.

A few things in particular were not consistent before this PR:

- Empty list items in the middle of a list would not get rendered.
- Writing text and placing an embed on the next direct line would result in an extra newline on top of the embed.
- A line break was being place immediately at the end of InlineEmbeds (@mentions and emojis).

## Containerization

- The Parser now registers all formats and blots instead of having them hardcoded.
- All default Formats and Blots are registered in the bootstrap from a single call (`addCoreBlotsAndFormats()`)
- I've updated the embed manager to be able to register all of its "core" embeds at once as well (`addCoreEmbeds()`). 10-15 registration methods don't need to be duplicated between test bootstrap and the main bootstrap.

## Parsing structure

In order to get more consistent rendering, particularly around newlines (some extra were getting inserted all over the place), I had to make the parsing much more robust. Before the parser was pretty much an afterthought and had no tests so it wasn't really parsing strictly enough. 

It was actually creating a lot of extra Blots and BlotGroups with newlines, and there was some numerous checks in the renderer of whether or not we should skip rendering a certain blot. 

**Before**

The parser had 2 steps:

- `splitPlainTextNewlines` - Splits plain text operations into multiple operations on their newlines and insert some break markers.
- `parse` - Create blots out of the the break markers and sort them into groups. Sometimes blots would create additional break markers when they were created.

And rendering was:

- `render` - Iterate through the groups. Because we were parsing extra newlines, there was some logic in this step to ignore rendering certain items. It was complicated and fragile.

**After**

The parser has 3 steps:

- `splitPlainTextNewlines` - Splits plain text operations into multiple operations on their newlines. This 100% recreates the array of operations.
- `insertBreakPoints` - Loop through the array of split operations and replace certain newlines (totally replace them, there text is no longer of any value) with a constant `BREAK_OPERATION`.
- `createBlotGroups` - Organize the blots into their requisite BlotGroups. There are 3 ways to determine the end of a group:
  - A `BREAK_OPERATION` is encountered. This clears the previous group and moves on to the next operation.
  - A newly created blot returns `true` from `$blot->shouldClearCurrentGroup()`. This clears the previous group _before_ inserting the newly created blot.
  - A newly created blot returns `true` from `$blot->isOwnGroup()`. This clears the group _after_ inserting the newly created blot.

The rendering now only does one thing:

- Iterate through all of the groups and blots and call `render()` on them. Every `Blot` and `BlotGroup` that was parsed is now rendered.

## Testing

**The renderer**

We now test the rendered HTML by pre-parsing it a bit so that newlines don't affect us. I tried using htmLawed and other built in libraries, but they didn't produce consistent enough output on all of the HTML structures.

Now expectations and actual html results are compared by each going through the following process first:

- Striping 0 width whitespace.
- HtmlMin() - Uses an HTML minified to remove as much whitespace as possible.
- Stubbing out svg's with `<SVG />`. Inline svg's were taking up way to much space in the diffs.
- Use regex to insert a newline before every new tag. This doesn't lend itself to "beautiful" html, but it is consistent, and makes diffs on failed tests a lot easier to read than when there are no new lines.

**The parser**

I've added a method to parser for testing purposes only. I don't no if we have a naming convention for this but it is called `parseIntoTestData()`. Instead of return BlotGroups with Blots in side of them, it returns arrays with classnames inside of them. This way we can do simple equality assertions on the output.

For example:

A blot group with 3 blockquote lines in it would look like this:

```
[
    ["class" => BlockquoteLineBlot::class, "content" => "Line 1"],
    ["class" => BlockquoteLineBlot::class, "content" => "Line 3"],
    ["class" => BlockquoteLineBlot::class, "content" => "Line 7"],
],
```

### Simplified Blots

Blots no longer do:
- Any handling of break markers.
- Mostly any handling of newlines. (The can get replaced with a `<br>` but thats generally it) _*_

The following blots have been collapsing into their parent class because they only 1 child class:

- `AbstractBlockBlot` -> Collapsed into `TextBlot`
- `AbstractListBlot`, `BulletListBlot`, `OrderedListBlot` -> Collapsed into a single `ListLineBlot` that extends from the existing `AbstractLineBlot`. I found the logic of the how the list blots were assembled to be identical the other lines blots.
- `AbstractBlockEmbedBlot` -> Merged into `ExternalBlot`. That abstract blot was created before we had the embed manager, when I envisioned the different types of external embeds having their own blot.
- `AbstractFormat` no longer extends `AbstractBlot`. It's not a blot so it shouldn't extend that.

Additionally:

- The handling of `Format`s has been moved out from `TextBlot` into a `FormattableTextTrait`
- A `NullBlot` has been created. When a match for an operation can't be found, it will fall back to a text blot if it is a plain text insert. If its not (say a badly formatted embed blot), we will fallback to the `NullBlot` which renders nothing.
